### PR TITLE
Integrate with sbt language server

### LINKED ISSUE #169: Integrate with sbt v1.1 server
See announcement: https://developer.lightbend.com/blog/2017-11-30-sbt-1-1-0-RC1-sbt-server/index.html

It would be fun to try and detect when sbt server is running and send sbt/exec notifications on buffer save and forward diagnostics to the client. Might also be interesting to try and replace scalametaEnableCompletions.

A feature I'm personally quite interested in is test on save, I use `~foo/testOnly com.blah.Foo` all the time. It would be nice if the editor could instead trigger tests without using file watching.

### PR DESCRIPTION
**Pull request by olafurpg**
Fixes #169. 

This PR adds an integration to connect to sbt 1.1 language server
to compile via sbt on file save and publish error messages as
diagnostics. We reuse the LanguageServer and LanguageClient
utilities but we are now a client instead of server, compared to when
connecting to editors.

This PR adds a new SbtConnect command to re-estabilish a connection
with sbt in case of a disconnection, for example if you exit sbt and
open again. Ideally, we can figure this out automatically in the future,
but a manual approach is a good start.

There are two new settings to control this new feature

- sbt.enabled, must be true to use this feature, false by default.
- sbt.command, which command to execute on file save. This makes it
  possible to for example run or test instead of compile or select
  a subproject in a multi module build.

I did not upgrade the build to sbt 1.1 because intellij import doesn't
work with sbt 1.1 yet for some reason. Instead, I added a new subproject `test-workspace-sbt-1.1` to play around with this new feature.

There are no integration tests yet since this is quite hard to test. I hope that we add tests once https://github.com/sbt/ipcsocket is published.

### CODE REVIEW

**Comment by olafurpg** [comment]
BTW if you customize the command to `run` instead of `compile` the output will get logged in the sbt shell, not the editor.

**Comment by dwijnand** [comment]
> The first commit contains ~1k loc of copy-pasted code that we can throw away as soon as https://github.com/sbt/ipcsocket is published as a library.

> I hope that we add tests once https://github.com/sbt/ipcsocket is published.

Done: http://repo1.maven.org/maven2/org/scala-sbt/ipcsocket/ipcsocket/1.0.0/

**Comment by olafurpg** [comment]
Thanks @dwijnand I've replaced the first commit with a dependency on ipcsocket!

I cleaned up the implementation a bit with docstrings and integration tests. It's ready for review 😄

**Comment by olafurpg** [comment]
Hmm, seems like my integration tests don't work on travis. They run fine locally, any idea what might be happening?

**Comment by olafurpg** [comment]
A warning is now displayed when the sbt server disconnect, for example when the shell is exited

screen shot 2018-01-09 at 00 28 32

**Comment by olafurpg** [comment]
Hmm, maybe the integration test exits the sbt shell on input stream close 🤔  We need to keep the process running until we establish the connection, will try with that tomorrow

**Comment by gabro** [comment]
@olafurpg quick question before I go through a more detailed review: does the current implementation support mixing scalafix/scalac diagnostics with sbt ones? At a glance of the code, it seems like they will be overriding each other.

**Comment by olafurpg** [comment]
> does the current implementation support mixing scalafix/scalac diagnostics with sbt ones?

No it doesn't. My idea was that we should either be using scalac for diagnostics or sbt, but never both. Scalafix diagnostics can work fine together with sbt server diagnostics, since scalafix can report errors after a successful compilation.

**Comment by gabro** [comment]
I may be misunderstanding, but due to how `publishDiagnostic` works, I think scalafix diagnostics and sbt diagnostics are currently tramping over each other.

- user saves
- sbt server compiles and sends diagnostics [A, B, C] back to metaserver
- metaserver sends sbt diagnostics [A, B, C] to the client
- scalafix runs on metaserver and sends diagnostics [D, E, F] to the client

Final result: the client displays [D, E, F]. If scalafix runs before sbt, then the client displays [A, B, C].
The client never displays [A, B, C, D, E, F], which is the desired outcome.

In short: metaserver must merge all the sources of diagnostics before sending them back to the client, because the client displays all and only the latest diagnostics received.

Am I getting this wrong?

**Comment by olafurpg** [comment]
We need to implement the configuration to disable the presentation compiler + scalafix on didChange for this to work. I should probably do that in this PR, I've so far only tested sbt server without scalametaEnableCompletions. If there is a compile error then no semanticdb file is generated and the compile errors from sbt are reported. On compile success, a semanticdb file is generated and scalafix will publish diagnostics.

**Comment by gabro** [comment]
Ah I see what you mean now: the very nature of those two sources (scalafix on compile and compile errors) is mutually exclusive.

Still I think we shouldn't rely on this: we may want to disable the presentation compiler errors (and rely on sbt instead), but at the same time keep using the PC for producing scalafix errors.

In general I find scalafix errors on a non-compiling buffer super interesting as a feature, and I wouldn't want to see this become mutually exclusive with the sbt integration.

I suggest we merge the sbt diagnostics with all the other diagnostics we produce, intead of blindly proxying them to the client. This should make us flexibly enough to support several mix&match scenarios.

**Comment by olafurpg** [comment]
We probably should add a specific data structure to keep track of the diagnostics from each source in the server side, when one source changes (sbt, scalac or scalafix) we can publish the new total list of diagnostics. I think this is how the java ls works. Do you want this to be implemented in this PR or as a follow up? I would personally prefer to get this in before it grows too big, by default this sbt server integration is disabled.

**Comment by gabro** [comment]
> We probably should add a specific data structure to keep track of the diagnostics from each source in the server side

I agree, since we start having multiple sources, with potentially different response time, it makes sense to make our server stateful to incrementally provider results.

Provided this PR doesn't break master (and it shouldn't, since it's disabled) I'm ok with working on it on a follow up. I'll try to review this asap so that we can merge this! :shipit:

**Comment by olafurpg** [comment]
I upgraded our build to sbt 1.1 and this made it significantly easier to test the server integration since we can query our build instead of using ProcessBuilder. It would be great to merge this PR asap to avoid conflicts with #174

**Comment by olafurpg** [comment]
It seems sbt server does not start on Travis https://travis-ci.org/scalameta/language-server/jobs/328669121#L505 The sbt version is correct but it never reports what socket it's listening to. Any ideas what might be wrong?

File: jsonrpc/src/main/scala/org/langmeta/jsonrpc/MessageWriter.scala [hunk 1]
```diff
@@ -34,13 +35,16 @@ class MessageWriter(out: OutputStream) extends LazyLogging {
     lock.synchronized {
       require(h.get(ContentLen).isEmpty)
 
-      val str = msg.asJson.noSpaces
+      val json = msg.asJson.withObject(
+        obj => Json.fromJsonObject(obj.add("jsonrpc", Json.fromString("2.0")))
+      )
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply jsonrpc/src/main/scala/org/langmeta/jsonrpc/MessageWriter.scala]: Nice! Didn't know about `"2.0".asJson`. I went with 

```
val json = msg.asJson.withObject(_.add("jsonrpc", "2.0".asJson).asJson)
```

since we only want to add this if it's an object.

I find it a bit annoying to require `.asJson` on JsonObject, do you have an intuition why it's like this?

**Comment by olafurpg** [comment]
`startServer` seemed to do the trick! Looks like this is finally ready for merge 😄 

PS. quite happy with scalafixTest error messages

```diff
--- /home/travis/build/scalameta/language-server/jsonrpc/src/main/scala/org/langmeta/jsonrpc/MessageWriter.scala
+++ <expected fix from ExplicitResultTypes+RemoveUnusedImports+Disable>
@@ -4,7 +4,6 @@
 import java.nio.charset.StandardCharsets
 import com.typesafe.scalalogging.Logger
 import io.circe.Encoder
-import io.circe.Json
 import io.circe.syntax._
```

File: jsonrpc/src/main/scala/org/langmeta/jsonrpc/MessageWriter.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabro) [reply jsonrpc/src/main/scala/org/langmeta/jsonrpc/MessageWriter.scala]: I guess JsonObject is a complex data structure that it's unlreated to the Json AST per se. Anyway, I'm from the phone so I can't try, but I think mapObject would do the trick, since it's from JSONObject => JSONObject
